## issue
## 对2，4，6三个文件进行如下统计
1、对每一个文件根据指定参数统计cluster的个数
2、统计出不同的gene name 在不同的文件中的不重复cluster的个数，并合并为一个文件
方案如下：·
wt2,4,6
    对2，4，6 文件按如下参数分片cluster
    coverage:20,frequency:1,record:3,threashold:2500
        ``` 
            cd ~/GitHub/biotools.ucsc/genome/tools
            ./app.py -c ../data/20170617 -t 2500 -o 20 -e 3 -f 1
        ```
        每个基因有多少cluster（不重复）,生成如下格式
            geneName,WT2-Cluster-Num,WT4-Cluster-Num,WT6-Cluster-Num
        ```
            cd ~/excel2mysql/tools/docker-mysql
            make all
            phpmyadmin->http://localhost:8081 (make sure bio database exist)
            cp -a ~/GitHub/biotools.ucsc/genome/result/20170617 ~/GitHub/excel2mysql/data/
            cd ~/excel2mysql/tools/
            ./app.py -s settings.ini -b import -i ../data/20170617/ -o 1 -t tbl20170617
        ```
        ```
            open http://localhost:8081

            ## 查看一共有多少不同的gene name 分布在三个文件中
            SELECT DISTINCT `gene_name_(Mus_musculus_ensembl_v86_Genes)` as geneName FROM `tbl20170617`

            ## 获取三个文件中一共有多少不同的geneName 14638
            SELECT DISTINCT `gene_name_(Mus_musculus_ensembl_v86_Genes)` as gene_name FROM `tbl20170617` WHERE `gene_name_(Mus_musculus_ensembl_v86_Genes)` != '' ORDER BY `gene_name_(Mus_musculus_ensembl_v86_Genes)` ASC
            ## 基于上述 sql 新建视图 distinct_genename

            ## 统计 wt2 中每个gene_name中不同的cluster 个数
            SELECT Count(DISTINCT ClusterName) AS cluster_number,`gene_name_(Mus_musculus_ensembl_v86_Genes)` as gene_name,filename FROM `tbl20170617` WHERE filename LIKE '%WT2' AND `gene_name_(Mus_musculus_ensembl_v86_Genes)` != '' AND ClusterName !='' GROUP BY `gene_name_(Mus_musculus_ensembl_v86_Genes)` ORDER BY Chromosome ASC, Region ASC
            ## 基于上述sql 新建视图 wt2_genename_cluster

            #### 同上，构建wt4_genename_cluster，wt6_genename_cluster 2个视图

            ## 把wt2_genename_cluster的结果map到全局distinct_genename中去
            SELECT dg.gene_name,w2gc.cluster_number,w2gc.filename FROM distinct_genename dg LEFT JOIN wt2_genename_cluster w2gc ON dg.gene_name=w2gc.gene_name ORDER BY dg.gene_name ASC
            ## 基于上述sql，生成视图wt2_all_genename_cluster
            ## 同理 生成wt4_all_genename_cluster，wt6_all_genename_cluster
            ## 手工导出上述三个视图为csv即可，然后手工合并为一个文件（保持默认排序）
            #### 
            ## 备份
            ```
                 cd ~/excel2mysql/tools/docker-mysql
                 make sync
                 cp container-data/mariadb/INITDB/bio.sql ~/GitHub/biotools.ucsc/genome/task/20170617/
            ```



DELIMITER $$

DROP PROCEDURE IF EXISTS `my_proc` $$
CREATE PROCEDURE `my_proc`(arg1 INT) -- 1 input argument; you might need more or fewer
BEGIN

-- declare the program variables where we'll hold the values we're sending into the procedure;
-- declare as many of them as there are input arguments to the second procedure,
-- with appropriate data types.

DECLARE val1 INT DEFAULT NULL;
DECLARE val2 INT DEFAULT NULL;

-- we need a boolean variable to tell us when the cursor is out of data

DECLARE done TINYINT DEFAULT FALSE;

-- declare a cursor to select the desired columns from the desired source table1
-- the input argument (which you might or might not need) is used in this example for row selection

DECLARE cursor1 -- cursor1 is an arbitrary label, an identifier for the cursor
 CURSOR FOR
 SELECT t1.c1, 
        t1.c2
   FROM table1 t1
  WHERE c3 = arg1; 

-- this fancy spacing is of course not required; all of this could go on the same line.

-- a cursor that runs out of data throws an exception; we need to catch this.
-- when the NOT FOUND condition fires, "done" -- which defaults to FALSE -- will be set to true,
-- and since this is a CONTINUE handler, execution continues with the next statement.   

DECLARE CONTINUE HANDLER FOR NOT FOUND SET done = TRUE;

-- open the cursor

OPEN cursor1;

my_loop: -- loops have to have an arbitrary label; it's used to leave the loop
LOOP

  -- read the values from the next row that is available in the cursor

  FETCH NEXT FROM cursor1 INTO val1, val2;

  IF done THEN -- this will be true when we are out of rows to read, so we go to the statement after END LOOP.
    LEAVE my_loop; 
  ELSE -- val1 and val2 will be the next values from c1 and c2 in table t1, 
       -- so now we call the procedure with them for this "row"
    CALL the_other_procedure(val1,val2);
    -- maybe do more stuff here
  END IF;
END LOOP;

-- execution continues here when LEAVE my_loop is encountered;
-- you might have more things you want to do here

END $$

DELIMITER ;
            